src/collector.rs

```diff
@@ -0,0 +1,77 @@
+use std::{
+    alloc::{alloc, Layout},
+    cell::{Cell, UnsafeCell},
+    collections::HashMap,
+    path::Path
+};
+
+use crate::{
+    safepoints::{ReturnAddress, SafepointRoots},
+    Scan
+};
+
+/// The size of the heap in bytes
+const HSIZE: usize = 1024;
+
+/// The byte alignment of the heap
+const HALIGN: usize = 4;
+
+pub(crate) struct Collector {
+    hptr: Cell<*mut usize>,
+    hstart: Cell<usize>,
+    hend: Cell<usize>,
+
+    collect_next: Cell<bool>,
+
+    roots: UnsafeCell<Option<HashMap<ReturnAddress, SafepointRoots>>>
+}
+
+impl Collector {
+    pub(crate) fn new() -> Self {
+        Collector {
+            hptr: Cell::new(0 as *mut usize),
+            hstart: Cell::new(0),
+            hend: Cell::new(0),
+
+            collect_next: Cell::new(false),
+            roots: UnsafeCell::new(None)
+        }
+    }
+
+    #[inline]
+    pub fn collect_next(&self) {
```

Comment:
I'm not sure why, but this name doesn't really have the right sense to me. It's more `please_collect_soon` or similar!

Reply:
What `collect_next` really means is `this_will_collect_the_next_time_we_hit_a_safepoint_poll`. "Please" implies that there may be a reason that the future collection may be aborted/postponed, but at the moment, that's not the case.